src/Servers/Kestrel/Core/test/KestrelServerTests.cs

```diff
@@ -145,6 +145,64 @@ public void StartWarnsWhenIgnoringIServerAddressesFeature(string ignoredAddress)
         }
     }
 
+    [Theory]
+    [InlineData(null)] // Uses the default
+    [InlineData(HttpProtocols.Http1)]
+    [InlineData(HttpProtocols.Http2)]
+    public void NoTlsLogging_None(HttpProtocols? protocols)
+    {
+        var (warnings, infos) = GetNoTlsLogging(protocols);
+        Assert.Empty(warnings);
+        Assert.Empty(infos);
+    }
+
+    [Theory]
+    [InlineData(HttpProtocols.Http1 | HttpProtocols.Http3)]
+    public void NoTlsLogging_Info(HttpProtocols? protocols)
+    {
+        var (warnings, infos) = GetNoTlsLogging(protocols);
+        Assert.Empty(warnings);
+        Assert.Equal(2, infos.Count()); // ipv4 and ipv6
+    }
+
+    [Theory]
+    [InlineData(HttpProtocols.Http1AndHttp2)]
+    public void NoTlsLogging_Warn(HttpProtocols? protocols)
+    {
+        var (warnings, infos) = GetNoTlsLogging(protocols);
+        Assert.Equal(2, warnings.Count()); // ipv4 and ipv6
+        Assert.Empty(infos);
+    }
+
+    [Theory]
+    [InlineData(HttpProtocols.Http1AndHttp2AndHttp3)]
+    public void NoTlsLogging_WarnAndInfo(HttpProtocols? protocols)
+    {
+        var (warnings, infos) = GetNoTlsLogging(protocols);
+        Assert.Equal(2, warnings.Count()); // ipv4 and ipv6
+        Assert.Equal(2, infos.Count()); // ipv4 and ipv6
+    }
+
+    private static (IEnumerable<TestApplicationErrorLogger.LogMessage> warnings, IEnumerable<TestApplicationErrorLogger.LogMessage> infos) GetNoTlsLogging(HttpProtocols? protocols)
+    {
+        var testLogger = new TestApplicationErrorLogger();
+        var kestrelOptions = new KestrelServerOptions();
+
+        if (protocols.HasValue)
+        {
+            kestrelOptions.ConfigureEndpointDefaults(opt =>
+            {
+                opt.Protocols = protocols.Value;
+            });
+        }
+
+        using (var server = CreateServer(kestrelOptions, testLogger))
+        {
+            StartDummyApplication(server);
+            return (testLogger.Messages.Where(log => log.EventId == 64), testLogger.Messages.Where(log => log.EventId == 65));
```

Comment:
I don't love these magic numbers (though they arguably can't change without breaking back-compat).  Would it be worthwhile to define constants?

Reply:
This is tolerable since it's test code and the test will fail immediately if it ever changes.